Ath 1795 - Navbar

EoinMcD opened the pull request:
# Description

- Adds a Navbar that is present on all pages
- Buttons on the Navbar redirect to home page, and registration screen.
- When a user is logged in, the "login" and "register" buttons are hidden, and a "logout" button is shown
- When a user is not logged in, the "logout" button is hidden, and the "login" and "register" buttons are shown

# Testing
- Testing can be found at test/react/navbar.test.tsx
- Run 'npm test'

# Screenshots
### When user is not logged in:
![Screenshot 2021-04-23 at 14 41 45](https://user-images.githubusercontent.com/48068135/115879970-21fda580-a442-11eb-94f3-4bc7e4586f1a.png)
### When user is logged in:
![Screenshot 2021-04-23 at 14 42 21](https://user-images.githubusercontent.com/48068135/115880007-2a55e080-a442-11eb-854c-520a8ee62958.png)

EoinMcD committed:
feat: Navbar setup

EoinMcD committed:
feat: Navbar Work

EoinMcD committed:
test: Added Navbar Testing

EoinMcD committed:
Merge branch 'master' into ath-1795

EoinMcD committed:
feat: Styling and reworking tests

EoinMcD committed:
refactor: Removed bloat and updated snapshop

meganduffy commented on app/views/main/index.html.erb:
```diff
@@ -1,3 +1,4 @@
 
+<h1>MAIN BLOG PAGE</h1>
 <%= react_component("hello_name" ,{name: @name}) %>
+
```
Looks like excessive lines were added to this file

meganduffy commented on spec/controllers/user_spec.rb:
```diff
@@ -23,7 +23,7 @@
       it 'Logs the user in after sign up' do
         post :create, params: user
-        expect(!!controller.current_user).to eq(true)
+        expect(!controller.current_user.nil?).to eq(true)
```
This line of code is confusing. What is a simpler way you could write this?

EoinMcD replied:
Not sure why I changed this from what it was before? It worked fine before, so I changed it back

meganduffy commented on app/controllers/main_controller.rb:
```diff
@@ -1,5 +1,6 @@
 class MainController < ApplicationController
   def index
     @name = current_user.first_name
+    @logged_in = !!current_user
```
`@logged_in` has been added to the main controller but is only used in the application view. Do you think it would make more sense to add it to the application controller?

Also, consider the truthy nature of ruby. Would the presence of `current_user` work just as well as an indicator of being logged in as a converted boolean? Something to think about.

EoinMcD replied:
I've been looking into this for a while, I see how its a better idea, but not 100% sure if the way I have it implemented is the way you were thinking

meganduffy commented on app/views/layouts/application.html.erb:
```diff
@@ -12,6 +12,7 @@
   </head>
 
   <body>
+  <%= react_component("my_navbar", {logged_in: @logged_in, regPath: signup_path,homePath: root_path}) %>
```
Missing a space after comma.

meganduffy commented on app/javascript/components/my_navbar.tsx:
```diff
@@ -0,0 +1,53 @@
+  const regPath = props.regPath;
+  const homePath = props.homePath;
+  return (
+    <AppBar position="static" className={classes.toolbar}>
+  <Toolbar >
```
Indentation is off here. Do you have Prettier working with tsx files? It will save a lot of hassle to have that set up :)

EoinMcD replied:
Used prettier on this now

meganduffy commented on app/javascript/components/my_navbar.tsx:
```diff
@@ -0,0 +1,53 @@
+
+function MyNavbar (props) {
+  const classes = useStyles();
+  const logged = props.logged_in;
```
This could be renamed to `logged` to make it clearer.

EoinMcD replied:
Renamed

meganduffy replied:
I meant `logged_in`, oops!

EoinMcD replied:
Ah okay I was wondering why you wanted the change, and when I changed it, it wanted camelcase, so I've fixed that anyway

meganduffy commented on app/javascript/components/my_navbar.tsx:
```diff
@@ -0,0 +1,53 @@
+    <Typography variant="h1" className={classes.title}>
+      <Button data-testid="HomeButton" href={homePath} color="inherit">BlogEngine</Button>
+    </Typography>
+     {logged ? null : <Button data-testid="RegButton" href={regPath} color="inherit">Register</Button>}
+     {logged ? <Button color="inherit">Logout</Button> : <Button color="inherit">Login</Button>}
```
Why are you setting the color to be inherit here?

EoinMcD replied:
The inherit colour for the button is white, when I try to set color="white" , it doesnt work, Ill look into this more

meganduffy commented on app/javascript/components/my_navbar.tsx:
```diff
@@ -0,0 +1,53 @@
+    <AppBar position="static" className={classes.toolbar}>
+  <Toolbar >
+    <Typography variant="h1" className={classes.title}>
+      <Button data-testid="HomeButton" href={homePath} color="inherit">BlogEngine</Button>
```
data-testids should be written using kebab-case, just like CSS classes.

meganduffy commented on app/javascript/components/my_navbar.tsx:
```diff
@@ -0,0 +1,53 @@
+  logged_in: PropTypes.bool.isRequired,
+  regPath: PropTypes.string.isRequired,
+  homePath: PropTypes.string.isRequired
+
```
Extra newline.

meganduffy commented on test/react/navbar.test.tsx:
```diff
@@ -0,0 +1,52 @@
+  expect(container).toMatchSnapshot();
+});
+
+test("Title button has been passed the right link", () => {
```
When writing React tests, we would usually use a functional testing approach.
This would mean that we're not concerned with how the component achieves what we're testing, but instead focuses on the output. 

In that case, the test case should be something like `Home button has the correct link`.

If you want to read more about functional testing, this is a pretty good article:
https://medium.com/javascript-scene/behavior-driven-development-bdd-and-functional-testing-62084ad7f1f2

EoinMcD replied:
My bad, changed now

meganduffy commented on test/react/navbar.test.tsx:
```diff
@@ -0,0 +1,52 @@
+  expect(getByTestId(container, "HomeButton")).toHaveProperty("href", "http://localhost/");
+});
+
+test("Register button has been passed the right link", () => {
```
Same comment as above, we want to test the "what" instead of the "how".

meganduffy commented on test/react/navbar.test.tsx:
```diff
@@ -0,0 +1,52 @@
+
+test("Title button has been passed the right link", () => {
+  const { container } = renderNavbar();
+  expect(getByTestId(container, "HomeButton")).toHaveProperty("href", "http://localhost/");
```
This test will fail if your base URL changes, eg. if you pushed it to production. It's not a blocker at this time but can you think of any way to write this test so that it doesn't pose this issue?

EoinMcD replied:
I've tried looking into this but havent found any answers

meganduffy reviewed (commented):
Great work Eoin! One thing we could discuss in more depth is `describe` and `it` blocks for structuring your react tests. Let's follow up on that!

EoinMcD committed:
refactor: Worked on feedback from pr

Updated wording in testing, removed whitespace, moved code to more suitable places

EoinMcD committed:
feat: Addressed more feedback

Added newlines, fixed tests, moved code to more suitable ares

meganduffy commented on test/react/navbar.test.tsx:
```diff
@@ -0,0 +1,51 @@
+
+test("Title button has the right link", () => {
+  const { container } = renderNavbar();
+  expect(getByTestId(container, "home-button")).toHaveProperty("href", window.location.href);
```
Hmmmm sneaky! This does come with it's own set of issues (what if you were clicking the navbar from the /signups page for instance!) but it definitely works for the time being. Nice work around.

EoinMcD replied:
I dont understand why the example you gave wouldn't work? From what I've learned is that `window.location.href` gives the url before the ' / ' 
So clicking from " localhost:3000/signup " would still return just "localhost:3000 " 
Am I wrong in thinking this?

meganduffy reviewed (approved):
Just on the topic of `color="inherit"`, you really should be setting color using CSS. 
Something to keep in mind for the future - inline html styling is normally discouraged!
